CIP-XXXX/README.md

```diff
@@ -0,0 +1,139 @@
+Next, we will define the query layer API.
+We attempt to define the API in a transport-agnostic way: the request and return types for each endpoint are as defined in CIP-0116, with a few notable exception where the corresponding CDDL type lacked some useful information (more on this can be found in the `Query Layer API` section). Finally, we end this section with some notes on rollbacks and pagination.
+
+#### Existing Query Layer designs
+
+There are two approaches to Cardano dApp development:
+
+1. **Using customized chain followers**. A chain follower is a program that interacts with cardano-node and processes all incoming transactions, as well as rollbacks, to maintain consistent dApp-specific state. Example: [Carp](https://dcspark.github.io/carp/docs/intro/).
+
+2. **Using general-purpose query layers**. General-purpose query layers allow to query blockchain data using a wide set of APIs that are not built with a particular dApp domain in mind. dApp state has to be constructed based on data returned from the queries. Examples: Blockfrost, Maestro.
+
+The first approach allows for lower runtime resource consumption, but a general-purpose query layer has an advantage of being more easily reusable between dApps.
+
+In this proposal, we are focusing on general-purpose querying only.
+
+### Query Layer API
+
+This section contains descriptions for methods & their parameter lists.
+
+The scope of this section is loosely based on a [comparison table for existing Cardano query layers](./Query_Layer_API_Comparison.md).
+The goal is to make it so that the API could be implemented via simple adapters that transform requests and responses to the appropriate formats.
+
+The payload formats used below are either references to [CIP-0116 - Standard JSON encoding for Domain Types](https://cips.cardano.org/cip/CIP-0116), which specifies cardano domain types via a JSON schema, or references to the [Query Layer JSON schema](./query-layer.json) which we defined in this CIP to define some types that are not present in the CDDL spec.
+
+[View the list of endpoints here](./endpoints.md)
+
+### Transports
+
+The API can be implemented across several transports. The goal is to allow several different clients, possibly written in different languages, to interact with wallets.
+For this reason we provide an [Openapi schema](./open-api.json), a [JSON-RPC](./json-rpc.json) schema, and an interface for an [injected Javascript](./ts-api.md) object in Typescript.
+We generate these interfaces from a high level specification of the endpoints [source](https://github.com/mlabs-haskell/query-layer-impl), ensuring that the information is consistent and easily updatable for different choices of transport layer.
+
+### Pagination
+
+In CIP-30, pagination is not reliable, because there is no guarantee that the set of UTxOs does not change between calls. This behavior is not suitable for DeFi: consistency should be prioritized, and pagination should be avoided.
```

Comment:
What are your thoughts on queries that result in large responses? I feel that this design decision to differ from CIP-0030 is a conscious one, but perhaps it would also be helpful to add some additional details on the downside of this choice.

Also note that by having rollbacks, consistency of data can never be 100% of the time.

Reply:
Thanks @perturbing that is a very good point! I added some more to this section, hopefully it better explains the tradeoff now.